docker_cli/commands.py

```diff
@@ -0,0 +1,143 @@
+    except ImageNotFoundError as err:
+        print(f"Error: {str(err)}")
+        exit(1)
+
+    image.drop_in()
+    print("Drop-in session ended.")
+    exit()
+
+
+def remove(
+    tags: Iterable[str],
+    force: bool = True,
+    quiet: bool = False,
+    ignore_prefix: bool = False,
+) -> None:
+    """
+    Remove all Docker images that fit a given tag or wildcard.
+
+    This tag or wildcard will take the form [UNIVERSAL PREFIX]-[tag or wildcard] if the
+    prefix does not already fit this.
+
+    Parameters
+    ----------
+    force : bool, optional
+        Whether or not to force the removal. Defaults to True.
+    tags : Iterable[str]
+        An iterable of tags or wildcards to be removed.
+    quiet : bool, optional
+        Whether or not to run quietly. Defaults to False.
+    ignore_prefix: bool, optional
+        Whether or not to ignore the universal prefix and only use the tag or wildcard.
+        Use with caution, as this will remove ALL images matching the wildcard.
+        e.g. remove(["*"], ignore_prefix = True) will remove all images.
+    """
+    force_arg = "--force " if force else ""
+    output = DEVNULL if quiet else None
+
+    # Search for and delete all images fitting each tag or wildcard.
+    for tag in tags:
+        prefix = universal_tag_prefix()
+        search = tag if (tag.startswith(prefix) or ignore_prefix) else f"{prefix}-{tag}"
+        if not quiet:
+            print(f"Attempting removal for tag: {search}")
+
+        # Search for all images whose name fits this tag, acquire a list
+        search_command = split(f'docker images --filter=reference="{search}" -q')
+        try:
+            search_result = run(search_command, text=True, stdout=PIPE, stderr=output)
+        # A CalledProcessError here indicates that the search failed. Skip to the next.
+        except CalledProcessError as err:
+            if not quiet:
+                print(
+                    f"Search for {search} failed with error code {err.returncode}. "
+                    "Proceeding."
+                )
+            continue
+        # An empty return indicates that no such images were found. Skip to the next.
+        if search_result.stdout == "":
+            if not quiet:
+                print(f"No images found fitting pattern {search}. Proceeding.")
+            continue
+        # The names come in a list delimited by newlines. Reform this to be delimited
+        # by spaces to use with `Docker rmi`.
+        search_result_str = " ".join(search_result.stdout.split("\n"))
+
+        # Remove all images in the list
+        command = split(f"docker rmi {force_arg}{search_result_str}")
+        command += []
+        try:
+            run(command, stdout=output, stderr=output)
+        except CalledProcessError as err:
+            if not quiet:
+                print(
+                    f"Removal of {tag} failed with error code {err.returncode}. "
+                    "Proceeding."
+                )
```

Comment:
Responding to this and the above - there are a number of occasions (e.g. multiple tags pointing to the same image, a prior search already having deleted all the images pointed to by this one) that a prior removal in the series of those deleted by this command may have already deleted this image. When this happens, the command will fail to find and/or delete some images because they already don't exist.

This happens with some regularity, so I figured it should just skip over errors like this since the command is typically operating as intended (i.e. all requested images are deleted even if some of the subcommands throw errors, no unrequested images are deleted) in this case.

Reply:
Since it's difficult to run the command without errors being raised even during normal operation, maybe it'd be better to switch `quiet` to `verbose` and only print these error messages if they're requested?

Reply:
>multiple tags pointing to the same image

Hmm, not sure I understand why this would cause an error from either the `docker images` or `docker rmi` steps.

>a prior search already having deleted all the images pointed to by this one

If all the images that matched a particular search query had previously been deleted, wouldn't the search results then just be an empty set? It looks like that case is currently being handled here:

https://github.com/isce3-testing/image-sandbox/blob/5c767e0dc4f463b0a02311d887a1a66bf2ed7af4/docker_cli/commands.py#L72-L76

So I don't expect that that scenario would cause an error from the `docker images` or `docker rmi` commands either.

Am I missing something? Is there a set of commands that I could try running in order to reproduce an error here?

If possible, I think we should try to address specific error cases before resorting to treating all errors from these steps as non-fatal.

Reply:
I was trying to run through the process to recreate the effect and realized that I had already fixed the problem these were meant to address! Removing this logic now.